linera-views/src/views/key_value_store_view.rs

```diff
@@ -305,61 +312,71 @@ impl<C: Context> View for KeyValueStoreView<C> {
                         .context
                         .base_key()
-                        .base_tag_index(KeyTag::Index as u8, &index);
-                    batch.put_key_value_bytes(key, value);
+                        .base_tag_index(KeyTag::Index as u8, index);
+                    batch.put_key_value_bytes(key, value.clone());
                     delete_view = false;
                 }
             }
-            self.stored_hash = None
         } else {
-            for index in mem::take(&mut self.deletion_set.deleted_prefixes) {
+            for index in self.deletion_set.deleted_prefixes.iter() {
                 let key = self
                     .context
                     .base_key()
-                    .base_tag_index(KeyTag::Index as u8, &index);
+                    .base_tag_index(KeyTag::Index as u8, index);
                 batch.delete_key_prefix(key);
             }
-            for (index, update) in mem::take(&mut self.updates) {
+            for (index, update) in self.updates.iter() {
                 let key = self
                     .context
                     .base_key()
-                    .base_tag_index(KeyTag::Index as u8, &index);
+                    .base_tag_index(KeyTag::Index as u8, index);
                 match update {
                     Update::Removed => batch.delete_key(key),
-                    Update::Set(value) => batch.put_key_value_bytes(key, value),
+                    Update::Set(value) => batch.put_key_value_bytes(key, value.clone()),
                 }
             }
         }
-        self.sizes.flush(batch)?;
-        let hash = *self.hash.get_mut().unwrap();
+        self.sizes.pre_save(batch)?;
+        let hash = *self.hash.lock().unwrap();
         if self.stored_hash != hash {
             let key = self.context.base_key().base_tag(KeyTag::Hash as u8);
             match hash {
                 None => batch.delete_key(key),
                 Some(hash) => batch.put_key_value(key, &hash)?,
             }
-            self.stored_hash = hash;
         }
         if self.stored_total_size != self.total_size {
             let key = self.context.base_key().base_tag(KeyTag::TotalSize as u8);
             batch.put_key_value(key, &self.total_size)?;
-            self.stored_total_size = self.total_size;
         }
-        self.deletion_set.delete_storage_first = false;
         Ok(delete_view)
     }
 
+    fn post_save(&mut self) -> Result<(), ViewError> {
+        if self.deletion_set.delete_storage_first {
+            self.stored_total_size = SizeData::default();
+            self.stored_hash = None;
+        }
+        self.deletion_set.delete_storage_first = false;
+        self.deletion_set.deleted_prefixes.clear();
+        self.updates.clear();
+        self.sizes.post_save()?;
+        let hash = *self.hash.lock().unwrap();
+        self.stored_hash = hash;
+        self.stored_total_size = self.total_size;
+        Ok(())
+    }
```

Comment:
This still looks wrong. (The lines `if self.deletion_set.delete_storage_first { .. }` do nothing)

Reply:
Perhaps the original code is wrong